dev/PushNotifications/PushNotificationManager.cpp

```diff
@@ -0,0 +1,263 @@
+            // Returns a com_ptr to the implementation type
+            auto pushChannel =
+                winrt::make_self<winrt::Microsoft::Windows::PushNotifications::implementation::PushNotificationChannel>(pushChannelReceived);
+
+            channelResult = winrt::make<winrt::Microsoft::Windows::PushNotifications::implementation::PushNotificationCreateChannelResult>(
+                *(pushChannel.get()),
+                channelRequestResult,
+                status);
+        }
+        else if (status == PushNotificationChannelStatus::CompletedFailure)
+        {
+            channelResult = winrt::make<winrt::Microsoft::Windows::PushNotifications::implementation::PushNotificationCreateChannelResult>(
+                nullptr,
+                channelRequestResult,
+                status);
+        }
+
+        co_return channelResult;
+    }
+
+    PushNotificationRegistrationToken PushNotificationManager::RegisterActivator(PushNotificationActivationInfo const& details)
+    {
+        winrt::guid taskClsid = details.TaskClsid();
+        DWORD cookie = 0;
+        BackgroundTaskRegistration registeredTask = nullptr;
+
+        if (WI_IsFlagSet(details.Kind(), PushNotificationRegistrationKind::PushTrigger))
+        {
+            auto tasks = BackgroundTaskRegistration::AllTasks();
+            bool taskRegistered = std::any_of(begin(tasks), end(tasks), [&](auto&& task) { return task.Value().Name() == backgroundTaskName; });
+
+            if (!taskRegistered)
+            {
+                BackgroundTaskBuilder builder;
+                builder.Name(backgroundTaskName);
+
+                PushNotificationTrigger trigger{};
+                builder.SetTrigger(trigger);
+
+                if (taskClsid != winrt::guid())
+                {
+                    auto builder5 = builder.try_as<winrt::IBackgroundTaskBuilder5>();
+                    if (IsPackagedProcess() && builder5)
+                    {
+                        builder5.SetTaskEntryPointClsid(taskClsid);
+                        winrt::com_array<winrt::IBackgroundCondition> conditions = details.GetConditions();
+                        for (auto condition : conditions)
+                        {
+                            builder.AddCondition(condition);
+                        }
+                    }
+                    else
+                    {
+                        throw winrt::hresult_not_implemented();
+                    }
+                }
+
+                registeredTask = builder.Register();
+            }
+        }
+
+        if (WI_IsFlagSet(details.Kind(), PushNotificationRegistrationKind::ComActivator))
+        {
+            if (taskClsid != winrt::guid())
+            {
+                {
+                    auto lock = g_lock.lock();
+                    // Define handle that will be set during background task execution
+                    g_waitHandleForArgs = wil::unique_handle(CreateEvent(nullptr, FALSE, FALSE, nullptr));
```

Comment:
If this `CreateEvent` fails, or if the `CoRegisterClassObject` below fails, or if the attempt to construct the `PushNotificationRegistrationToken` fails (out of memory, say), we fail to clean up: *the background task is still registered*. This can cause the app's internal state to get confused, such as our own unit test.